### ISSUE TITLE
something closes stdin when running tests on Windows, causing problems with test_scripts

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5891]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5891|
|Type|defect|
|Created|2012-08-20 01:42:08Z|
|Branch|https://github.com/twisted/twisted/tree/no-stdin-for-scripts-5891|

This probably happens on other platforms too, but for some reason I saw this failure when I set up my most recent Windows VM for testing (win7, x64).

```
[ERROR]
Traceback (most recent call last):
  File "C:\Users\glyph\Projects\Twisted\trunk\twisted\scripts\test\test_scripts.py", line 88, in test_twistdPathInsert
    output = getstdout(script, '-ny', 'bar.tac')
  File "C:\Users\glyph\Projects\Twisted\trunk\twisted\scripts\test\test_scripts.py", line 26, in getstdout
    stdout=PIPE, stderr=dn).stdout.read()
  File "C:\Python27\lib\subprocess.py", line 672, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
  File "C:\Python27\lib\subprocess.py", line 787, in _get_handles
    p2cread = self._make_inheritable(p2cread)
  File "C:\Python27\lib\subprocess.py", line 826, in _make_inheritable
    _subprocess.DUPLICATE_SAME_ACCESS)
exceptions.WindowsError: [Error 6] The handle is invalid

twisted.scripts.test.test_scripts.ScriptTests.test_twistdPathInsert
```

This only happens when running the whole suite, so it's a bug in some other test, but it happens because `test_scripts` incorrectly delegates trial's stdin to the processes it's spawning.  We should track down what's closing the stdin handle too, since that's a global state manipulation, but these tests should not be able to be hung by some random script asking for user input.

<details><summary>Searchable metadata</summary>

```
trac-id__5891 5891
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_no_stdin_for_scripts_5891 branches/no-stdin-for-scripts-5891
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__windows windows
time__1345426928000000 1345426928000000
changetime__1347883286000000 1347883286000000
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [35303]) Branching to 'no-stdin-for-scripts-5891'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

The change in the branch is to simply pass a null file as stdin; stderr is already similarly a null file.

I'll run tests for this later, when I'm not at a sprint, since it only happens in the full suite anyway.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @glyph|
|-|-|

 * Is there a reason to keep a global list of devnull files? We're leaking fd for no obvious reasons. It would even be nice if we could close them explicitly once Popen is called.

 * getstdout should have a docstring.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph removed owner|
|-|-|

Replying to [therve](#note_4):
>  * Is there a reason to keep a global list of devnull files? We're leaking fd for no obvious reasons. It would even be nice if we could close them explicitly once Popen is called.

No, sorry, this is debug code that should not have been put into review.

>  * getstdout should have a docstring.

Yes, and also that was a dumb name.  It's fixed.

Also, I [forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/no-stdin-for-scripts-5891), so the next reviewer can have a look at that.

> Thanks!

Thanks for the review.

-glyph

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @glyph|
|-|-|

Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `closed`|
|-|-|

(In [35714]) Merge no-stdin-for-scripts-5891.

Author: glyph

Reviewer: therve

Fixes: [#5891](https://github.com/twisted/twisted/issues/5891)

Consistently spawn scripts for testing with no stderr or stdin.  This fixes an issue where the tests can become unreliable when running the full suite, especially on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

fwiw: 

   1. The fewer uses of the `file` builtin, the less work there'll be porting to Python 3
   1. `with` is a nice way to make sure the files get closed (but I would not object to try/finally either)
